persistent-postgresql/test/JSONTest.hs

```diff
@@ -40,431 +42,654 @@ insert' :: (MonadIO m, PersistStoreWrite backend, BaseBackend backend ~ SqlBacke
+  return $ EdgeCaseKeys {..}
+
+globalKeys :: (Monad m, MonadIO m) => ReaderT SqlBackend m (GlobalKeys TestValue)
+globalKeys = do
+          nullK <- insert' Null
+          let nulls = [nullK]
+
+          boolTK <- insert' $ Bool True
+          boolFK <- insert' $ toJSON False
+          let bools = [boolTK, boolFK]
+
+          num0K <- insert' $ Number 0
+          num1K <- insert' $ Number 1
+          numBigK <- insert' $ toJSON (1234567890 :: Int)
+          numFloatK <- insert' $ Number 0.0
+          numSmallK <- insert' $ Number 0.0000000000000000123
+          numFloat2K <- insert' $ Number 1.5
+          -- numBigFloatK will turn into 9876543210.123457 because JSON
+          numBigFloatK <- insert' $ toJSON (9876543210.123456789 :: Double)
+          let numbers = [ num0K, num1K, numBigK, numFloatK
+                        , numSmallK, numFloat2K, numBigFloatK ]
+
+          strNullK <- insert' $ String ""
+          strObjK <- insert' $ String "{}"
+          strArrK <- insert' $ String "[]"
+          strAK <- insert' $ String "a"
+          strTestK <- insert' $ toJSON ("testing" :: Text)
+          str2K <- insert' $ String "2"
+          strFloatK <- insert' $ String "0.45876"
+          let strings = [ strNullK, strObjK, strArrK, strAK
+                        , strTestK, str2K, strFloatK ]
+
+          arrNullK <- insert' $ Array $ V.fromList []
+          arrListK <- insert' $ toJSON [emptyArr,emptyArr,toJSON [emptyArr,emptyArr]]
+          arrList2K <- insert' $ toJSON [emptyArr,toJSON [Number 3,Bool False]
+                                        ,toJSON [emptyArr,toJSON [Object mempty]]
+                                        ]
+          arrFilledK <- insert' $ toJSON [Null, Number 4, String "b"
+                                         ,Object mempty, emptyArr
+                                         ,object [ "test" .= [Null], "test2" .= String "yes"]]
+          let lists = [ arrNullK, arrListK, arrList2K, arrFilledK ]
+
+          objNullK <- insert' $ Object mempty
+          objTestK <- insert' $ object ["test" .= Null, "test1" .= String "no"]
+          objDeepK <- insert' $ object ["c" .= Number 24.986, "foo" .= object ["deep1" .= Bool True]]
+          let objects = [ objNullK, objTestK, objDeepK ]
+          return GlobalKeys{..}
+
+
+data EdgeCaseKeys record =
+  EdgeCaseKeys { arrList3K  :: Key record
+               , arrList4K  :: Key record
+               , objEmptyK  :: Key record
+               , objFullK   :: Key record }
+
+data GlobalKeys record =
+  GlobalKeys { nulls   :: [Key record]
+             , bools   :: [Key record]
+             , numbers :: [Key record]
+             , strings :: [Key record]
+             , lists   :: [Key record]
+             , objects :: [Key record] }
+
 specs :: Spec
-specs = describe "postgresql's JSON operators behave" $ do
+specs =
+    describe "Testing JSON operators"
+    $ beforeAll setup
```

Comment:
So the 'magic' of `beforeAll` is that you can run a function `IO SomeDataType` and then the `SomeDataType` is provided to every `it`, like:

```haskell
beforeAll (pure "hello world") $ do
  it "gives us hello world" $ \string -> do
    string `shouldBe` "hello world"
```

So we can write `beforeAll (setup :: IO GlobalKeys)` and then each spec item can look like:

```
it "lmao" $ \GlobalKeys{..} -> ...
```

When you get to the edge cases, you can write `beforeAll` *again*, but this time, because it's under a `beforeAll (setup :: IO Thing)`, the type of the function `beforeAll (setupExtra :: Thing -> IO ExtraThing)`, and now the `it` specs under that would be 

```haskell
it "message" $ \ExtraThing {..} -> ...
```

This may be a cleaner way to structure this.

Reply:
Oh, I didn't know that about `beforeAll`, I like that way better!

Reply:
I am getting a type error when I try and nest a `beforeAll` beneath another `beforeAll`. I think it is because the `beforeAll` returns a `Spec`, but the do block that the nested `beforeAll` resides in expects a `SpecWith GlobalKeys`. I am sure I am missing something simple, but if you have any advice that would be appreciated!

Reply:
I am just going to make all the tests work with the same initial state. I think that will simplify everything.